fwcd/kotlin-language-server
# Setup JDT LS extension to provide Java + Kotlin interoperability

Fixes https://github.com/fwcd/kotlin-language-server/issues/192

Sets up a VSCode extension to allow Java code to see Kotlin code. The extension depends on both `vscode-java` and `vscode-kotlin`. It uses the JDT LS extension mechanism, through a custom JDT LS extension that provides kotlin project importers for maven and gradle.

This seems to work reasonably well for small projects at least. I don't know how well it performs on bigger projects, but I think the concept looks valid and we can always improve that later.

### Some notes:

- The VSCode extension uses the `javaExtensions` contribution point to provide `vscode-java` with the JAR for the JDT LS extension. The JDT LS uses the jar to find the kotlin project importers. The importers have higher priority than the default ones provided by the JDT LS itself.
- The JDT LS extension was built with Maven and Java since it would have been much harder to do it with Gradle and Kotlin (due to lack of documentation and editor support).
- The importers actually extend `MavenProjectImporter` and `GradleProjectImporter` provided by the JDT LS. As such, pretty much all the importing logic is done by the JDT LS. Once the core java project is imported (by `MavenProjectImporter` or `GradleProjectImporter`) we add the compiled kotlin files to the project's classpath. This is how the java code is then able to see the kotlin code.
- At the moment, the kotlin language server does not generate any bytecode (at least as far as I could tell). This PR also introduces this, since it is required for the JDT LS extension to work. The bytecode generated by the kotlin language server is written to a folder called `kls` at the root of the workspace. I chose this because the JDT LS messes with the default maven and gradle output paths (and we should try to isolate the language servers as much as possible). The code generation is only triggered on file save for performance reasons.
- Some language features are not supported yet (e.g., go to definition from java to kotlin doesn't work as you would expect it to), but the projects should no longer complain about missing symbols when referencing kotlin classes, etc.

### How to test this for now:

On the `vscode-java-kotlin` folder, run the `build.sh` script to build the JDT LS extension. Afterwards, run `npm install` to install the VSCode extension dependencies. Once this is done, you can either debug it (using F5 on VSCode), or package it and install it from a VSIX.
**Important**: The JDT LS extension will only work with the latest kotlin language server code (i.e, the one in this PR), so you should install the language server locally and setup your `vscode-kotlin` to use that (through the `kotlin.languageServer.path`configuration property)

This is very experimental (I cannot stress this enough). This is literally the first thing I wrote involving Eclipse PDE and the JDT LS API. With that said, I think the concept at least seems to work, so I'm guessing we can improve this more over time.

Feedback welcome :)
